src/Type/Php/RegexShapeMatcher.php

```diff
@@ -0,0 +1,111 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PHPStan\Analyser\TypeSpecifierContext;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\Constant\ConstantArrayTypeBuilder;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\IntegerRangeType;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use function array_key_last;
+use function array_keys;
+use function is_string;
+use function preg_match;
+use function preg_replace;
+use const PREG_OFFSET_CAPTURE;
+use const PREG_UNMATCHED_AS_NULL;
+
+final class RegexShapeMatcher
+{
+
+	/**
+	 * @param int-mask<PREG_OFFSET_CAPTURE|PREG_UNMATCHED_AS_NULL>|null $flags
+	 */
+	public function matchType(string $regex, ?int $flags, TypeSpecifierContext $context): Type
+	{
+		if ($flags !== null) {
+			$trickFlags = PREG_UNMATCHED_AS_NULL | $flags;
+		} else {
+			$trickFlags = PREG_UNMATCHED_AS_NULL;
+		}
+
+		// add one capturing group to the end so all capture group keys
+		// are present in the $matches
+		// see https://3v4l.org/sOXbn, https://3v4l.org/3SdDM
+		$regex = preg_replace('~.[a-z\s]*$~i', '|(?<phpstanNamedCaptureGroupLast>)$0', $regex);
+
+		if (
+			$regex === null
+			|| @preg_match($regex, '', $matches, $trickFlags) === false
+		) {
+			return new ArrayType(new MixedType(), new StringType());
+		}
+		unset($matches[array_key_last($matches)]);
+		unset($matches['phpstanNamedCaptureGroupLast']);
+
+		$builder = ConstantArrayTypeBuilder::createEmpty();
+		foreach (array_keys($matches) as $key) {
+			// atm we can't differentiate optional from mandatory groups based on the pattern.
+			// So we assume all are optional
+			$optional = true;
```

Comment:
anyone has ideas on how we could identify which groups are optional and which not?

with the current PRs implementation most offsets are treated as optional, which makes the inferred type pretty weak.
see e.g. https://github.com/phpstan/phpstan-src/pull/2589/files#diff-8a0ca036dacc90b9ccceaa1a9053a14f13a279e04687c683397085b0a3b7c573R9

Reply:
I think you will have to parse the full regex.

Reply:
@gharlan just pointed me to https://github.com/hoaproject/Regex/issues/38#issue-386915806 which seems to solve the problem in Hoa

Reply:
Great finding! The fix seems correct based on https://github.com/hoaproject/Compiler/blob/b675e53267/README.md?plain=1#L134.